sklearn/decomposition/_dict_learning.py

```diff
@@ -1844,39 +2226,108 @@ def fit(self, X, y=None):
+                FutureWarning,
+            )
+            self._batch_size = 3
 
-        U, (A, B), self.n_iter_ = dict_learning_online(
-            X,
-            self.n_components,
-            alpha=self.alpha,
-            n_iter=self.n_iter,
-            return_code=False,
-            method=self.fit_algorithm,
-            method_max_iter=self.transform_max_iter,
-            n_jobs=self.n_jobs,
-            dict_init=self.dict_init,
-            batch_size=self.batch_size,
-            shuffle=self.shuffle,
-            verbose=self.verbose,
-            random_state=random_state,
-            return_inner_stats=True,
-            return_n_iter=True,
-            positive_dict=self.positive_dict,
-            positive_code=self.positive_code,
+        X = self._validate_data(X, dtype=np.float64, order="C", copy=self.shuffle)
+
+        self._check_params(X)
+        self._random_state = check_random_state(self.random_state)
+        n_samples, n_features = X.shape
+
+        dictionary = self._initialize_dict(X, self._random_state)
+        dict_buffer = dictionary.copy()
+
+        if self.shuffle:
+            X_train = X.copy()
+            self._random_state.shuffle(X_train)
+        else:
+            X_train = X
+
+        if self.verbose:
+            print("[dict_learning]")
+
+        # Inner stats
+        self._inner_stats = (
+            np.zeros((self._n_components, self._n_components)),
+            np.zeros((n_features, self._n_components)),
         )
-        self.components_ = U
-        # Keep track of the state of the algorithm to be able to do
-        # some online fitting (partial_fit)
-        self.inner_stats_ = (A, B)
-        self.iter_offset_ = self.n_iter
-        self.random_state_ = random_state
+
+        if self.max_iter is not None:
+
+            # Attributes to monitor the convergence
+            self._ewa_cost = None
+            self._ewa_cost_min = None
+            self._no_improvement = 0
+
+            batches = gen_batches(n_samples, self._batch_size)
+            batches = itertools.cycle(batches)
+            n_steps_per_epoch = int(np.ceil(n_samples / self._batch_size))
+            n_steps = self.max_iter * n_steps_per_epoch
+
+            # allow max_iter = 0
+            i = -1
+            import time
+
+            t = 0
+            for i, batch in zip(range(n_steps), batches):
+                t1 = time.time()
+                this_X = X_train[batch]
+
+                batch_cost = self._minibatch_step(
+                    this_X, dictionary, self._random_state, i
+                )
+                t += time.time() - t1
+                if self._minibatch_convergence(
+                    this_X, batch_cost, dictionary, dict_buffer, n_samples, i, n_steps
+                ):
+                    break
+                print(f"[{self._ewa_cost},{t}],")
+
+                if self.callback is not None:
+                    self.callback(locals())
+
```

Comment:
This needs to be covered by a test.

Reply:
I had to introduce this param to make it possible that the function calls the class instead of the opposite since only the function had this param.

It was however not tested anywhere. And I don't think that's something we want to keep in the long term. We'd rather have a unified callback API.

Reply:
Can you please make that explicit with a `# XXX` comment?